node/impl/full/eth_trace.go

```diff
@@ -0,0 +1,339 @@
+		et.Msg.Method == builtin.MethodsEAM.CreateExternal {
+		log.Debugf("found CreateExternal call at height: %d", height)
+		trace.Action.Input, err = decodePayload(et.Msg.Params, et.Msg.ParamsCodec)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+
+		if et.MsgRct.ExitCode.IsSuccess() {
+			// ignore return value
+			trace.Result.Output = nil
+		} else {
+			// return value is the error message
+			trace.Result.Output, err = decodePayload(et.MsgRct.Return, et.MsgRct.ReturnCodec)
+			if err != nil {
+				return xerrors.Errorf("buildTraces: %w", err)
+			}
+		}
+
+		// treat this as a contract creation
+		trace.SetCallType("create")
+	} else {
+		trace.Action.Input, err = handleFilecoinMethodInput(et.Msg.Method, et.Msg.ParamsCodec, et.Msg.Params)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+		trace.Result.Output, err = handleFilecoinMethodOutput(et.MsgRct.ExitCode, et.MsgRct.ReturnCodec, et.MsgRct.Return)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+	}
+
+	// TODO: is it OK to check this here or is this only specific to certain edge case (evm to evm)?
+	if et.Msg.ReadOnly {
+		trace.SetCallType("staticcall")
+	}
+
+	// there are several edge cases thar require special handling when displaying the traces
+	if parent != nil {
+		// Handle Native actor creation
+		//
+		// Actor A calls to the init actor on method 2 and The init actor creates the target actor B then calls it on method 1
+		if parent.Action.FilecoinTo == builtin.InitActorAddr &&
+			parent.Action.FilecoinMethod == builtin.MethodsInit.Exec &&
+			et.Msg.Method == builtin.MethodConstructor {
+			log.Debugf("Native actor creation! method:%d, code:%s, height:%d", et.Msg.Method, et.Msg.CodeCid.String(), height)
+			parent.SetCallType("create")
+			parent.Action.To = et.Msg.To.String()
+			parent.Action.Input = []byte{0x0, 0x0, 0x0, 0xFE}
+			parent.Result.Output = nil
+
+			// there should never be any subcalls when creating a native actor
+			return nil
+		}
+
+		// Handle EVM contract creation
+		//
+		// To detect EVM contract creation we need to check for the following sequence of events:
+		//
+		// 1) EVM contract A calls the EAM (Ethereum Address Manager) on method 2 (create) or 3 (create2).
+		// 2) The EAM calls the init actor on method 3 (Exec4).
+		// 3) The init actor creates the target actor B then calls it on method 1.
+		if parent.Parent != nil {
+			calledCreateOnEAM := parent.Parent.Action.FilecoinTo == builtin.EthereumAddressManagerActorAddr &&
+				(parent.Parent.Action.FilecoinMethod == builtin.MethodsEAM.Create || parent.Parent.Action.FilecoinMethod == builtin.MethodsEAM.Create2)
+			eamCalledInitOnExec4 := parent.Action.FilecoinTo == builtin.InitActorAddr &&
+				parent.Action.FilecoinMethod == builtin.MethodsInit.Exec4
+			initCreatedActor := trace.Action.FilecoinMethod == builtin.MethodConstructor
+
+			if calledCreateOnEAM && eamCalledInitOnExec4 && initCreatedActor {
```

Comment:
What if one of these steps fail? We should still treat it as an attempted creation.

Reply:
We have a few cases:

1. Failures. I.e., we may fail halfway through this path for some reason and need to treat it as a failed to creation.
2. Resurrection! EAM "create" and/or "create2" may not call the init actor but may, instead call "resurrect" on an existing but dead EVM actor. Let's handle that later, but it's an important part of "EVM-style" code upgrades (which are _not_ the same as FVM code upgrades).

Reply:
Added a TODO for this